Add documentation for the cstruct-stubgen tool (#87)
fox-it/dissect-docs: docs/source/projects/dissect.cstruct/index.rst

Don't like the C-like definition syntax? Write your own syntax parser!

cstruct-stubgen
---------------

``cstruct-stubgen`` is a tool that generates Python stub files (``.pyi``) from ``dissect.cstruct`` definitions.

> These stubs are useful for improving IDE support, enabling type checking, and enhancing developer experience when working with cstructs with code completion.

Reviewer: ```suggestion
These stubs provide an enhanced developer experience in your IDE, enabling type checking and code completion.
```

   class _c_systemtime(__cs__.cstruct):
       class SYSTEMTIME(__cs__.Structure):
           wYear: _c_systemtime.uint16
           wMonth: _c_systemtime.uint16
           wDayOfWeek: _c_systemtime.uint16
           wDay: _c_systemtime.uint16
           wHour: _c_systemtime.uint16
           wMinute: _c_systemtime.uint16
           wSecond: _c_systemtime.uint16
           wMilliseconds: _c_systemtime.uint16
           @overload
           def __init__(self, wYear: _c_systemtime.uint16 | None = ..., wMonth: _c_systemtime.uint16 | None = ..., wDayOfWeek: _c_systemtime.uint16 | None = ..., wDay: _c_systemtime.uint16 | None = ..., wHour: _c_systemtime.uint16 | None = ..., wMinute: _c_systemtime.uint16 | None = ..., wSecond: _c_systemtime.uint16 | None = ..., wMilliseconds: _c_systemtime.uint16 | None = ...): ...
           @overload
           def __init__(self, fh: bytes | memoryview | bytearray | BinaryIO, /): ...

   # Technically `c_systemtime` is an instance of `_c_systemtime`, but then we can't use it in type hints
   c_systemtime: TypeAlias = _c_systemtime

And the autocompletion, with an IDE such as vscode, will look like:

.. image:: /images/autocompletion.png

> 

Reviewer: ```suggestion

Structure fields also have correct typing, enabling further type checking and code completion on compatible Python types:

```

And maybe add a screenshot where you have a code completion dropdown on an integer.

Author: I will add an image like that

   import dissect.cstruct as __cs__
   from typing_extensions import TypeAlias

   class _c_systemtime(__cs__.cstruct):
       class SYSTEMTIME(__cs__.Structure):
           wYear: _c_systemtime.uint16
           wMonth: _c_systemtime.uint16
           wDayOfWeek: _c_systemtime.uint16
           wDay: _c_systemtime.uint16
           wHour: _c_systemtime.uint16
           wMinute: _c_systemtime.uint16
           wSecond: _c_systemtime.uint16
           wMilliseconds: _c_systemtime.uint16
           @overload
           def __init__(self, wYear: _c_systemtime.uint16 | None = ..., wMonth: _c_systemtime.uint16 | None = ..., wDayOfWeek: _c_systemtime.uint16 | None = ..., wDay: _c_systemtime.uint16 | None = ..., wHour: _c_systemtime.uint16 | None = ..., wMinute: _c_systemtime.uint16 | None = ..., wSecond: _c_systemtime.uint16 | None = ..., wMilliseconds: _c_systemtime.uint16 | None = ...): ...
           @overload
           def __init__(self, fh: bytes | memoryview | bytearray | BinaryIO, /): ...

   # Technically `c_systemtime` is an instance of `_c_systemtime`, but then we can't use it in type hints
   c_systemtime: TypeAlias = _c_systemtime

> And the autocompletion, with an IDE such as vscode, will look like:

Reviewer: ```suggestion
And the autocompletion, with an IDE such as VS Code, will look like:
```

.. image:: /images/autocompletion.png

Structure fields also have correct typing, enabling further type checking and code completion on compatible Python types:

> .. image:: /images/autocompletion_attribute.png

Reviewer: The image doesn't match the text description. This image should belong together with the previous image.
